duo_client/paginator.py

```diff
@@ -0,0 +1,198 @@
+
+        return six.text_type(md) if isinstance(md, int) else md
+
+    def _request_with_retry(self, method, path, params):
+        """Make an HTTP request via the specified Duo client, and retry if the
+        API responds with a 429 status. Method will raise MaxRequestAttemptsReached
+        if the client gets too many 429 status codes.
+
+        :param method: The method for the HTTP call ("GET", "POST", etc...)
+        :param path: The path to make the request on ("/sample/route", "/data"...)
+        :param params: Parameters (dict) to use in the request
+        :return: The json data from the API
+        :rtype: dict
+        """
+
+        # Generator which yields sleep durations.
+        backoff_gen = self._backoff_func(
+            self._wait_const,
+            self._min_wait_seconds,
+            self._max_attempts
+        )
+
+        for _, wait_seconds in backoff_gen:
+            try:
+                http_response, http_resp_data = self._duo_client.api_call(
+                    method,
+                    path,
+                    params
+                )
+                return self._duo_client.parse_json_response_and_metadata(
+                    http_response, http_resp_data
+                )
+            except RuntimeError as e:
+                if getattr(e, 'status', -1) in self._statuses_to_retry_on:
+                    sleep(wait_seconds)
+                else:
+                    raise e
+
+        raise MaxRequestAttemptsReached(
+            "Reached the max number of request attempts"
+        )
+
+    def request(
+        self,
+        method,
+        path,
+        params,
+        meta_offset_name=DEFAULT_META_OFFSET_NAME,
+        max_pages=DEFAULT_ALL_PAGES
+    ):
+        """Make a request on the specified API endpoint and return a generator.
+        The generator will make requests on the API and yield the pages.
+
+        :param method: The method for the HTTP call ("GET", "POST", etc...)
+        :param path: The path to make the request on ("/saf0ba235-0b33-23c8-bc23-a31aa0231de8ample/route", "/data"...)
+        :param params: Parameters (dict) to use in the request
+        :param meta_offset_name: The name of the subsequent param used to request the next page
+        :param max_pages: The number of pages to fetch (-1 means all pages)
+        :returns: A generator to yield pages from the api responses
+        :rtype: generator
+        """
+
+        while max_pages:
+            max_pages -= 1
+            response, metadata = self._request_with_retry(method, path, params)
+            next_offset = self._get_next_offset(response, metadata)
+
+            yield response
+
+            if next_offset:
+                params[meta_offset_name] = next_offset
+            else:
+                break
+
+    def request_all_pages(
+        self,
+        method,
+        path,
+        params,
+        meta_offset_name=DEFAULT_META_OFFSET_NAME,
+        max_pages=DEFAULT_ALL_PAGES
```

Comment:
if this function is called request_all_page then maybe it shouldn't take a max page argument?

Reply:
Yeah that's a good.  Maybe I should just completely remove this request_all_pages method?  It's just a simple abstraction over the request method and we can push down that responsibility to call next() on the pagination generator to the user. Does that sound okay?

Reply:
Yeah let's go with that for now. I think wrapping a generator in a list if you want all the values is a pretty common thing so we can let user code do that.